### PR TITLE
ASoC: SOF: Topology: Fix EQ feature by properly utilizing new user ABI

### PR DESCRIPTION
**Pull request by singalsu**
This patch contains a number of updates. First, the EQ header in uapi
is updated similarly as in firmware. The macros and structs were
conflicting and prevented firmware EQs to work correctly.

The topology.c sof_control_load_bytes() function is updated to check
the ABI fields instead of creating them. The user space tools need
to insert a valid ABI header into configuration data. The check for
user data size is improved by adding DSP IPC max size check.

The control.c ext bytes put/get code is similarly updated to
retrieve/return the ABI header from user space. The checks are improved
to cover size exceed and mismatch. The added debug messages and changed
error messages help to know the limits for user data size without need
to deep dive into source code.

The error message in snd_sof_bytes_put() function is fixed.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

### CODE REVIEW

**Comment by singalsu** [comment]
Note: The FW and topology fixes are needed too for the feature to work. Plus the tools fix is needed to be able to create working EQ configurations. I will create next the PRs for them.

I'm proposing to not bump the ABI version since I'm only fixing the buggy macros there.

File: sound/soc/sof/control.c [hunk 1]
```diff
@@ -303,13 +302,14 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
-		max_length = be->max;
-
-	dev_dbg(sdev->dev, "size of bytes put data is %d, max is %d\n",
-		header.length, max_length);
-	if (header.length > max_length) {
-		dev_err(sdev->dev, "error: size is too large\n");
+	dev_dbg(sdev->dev, "bytes: abi user data max size is %d\n", max_size);
+	dev_dbg(sdev->dev, "bytes: control bytes max is %d\n", be->max);
+	max_size = (be->max < max_size) ? be->max : max_size;
+	dev_dbg(sdev->dev, "bytes: size of bytes data is %d",
```
Comment (plbossart) [inline sound/soc/sof/control.c]: this seems too verbose to have a debug message every time you change an EQ coefficient.
Just keep the error message with all the information needed.

File: sound/soc/sof/control.c [hunk 2]
```diff
@@ -263,7 +263,7 @@ int snd_sof_bytes_put(struct snd_kcontrol *kcontrol,
 	pm_runtime_mark_last_busy(sdev->dev);
 	err = pm_runtime_put_autosuspend(sdev->dev);
 	if (err < 0)
-		dev_err(sdev->dev, "error: volume get failed to idle %d\n",
+		dev_err(sdev->dev, "error: bytes put failed to idle %d\n",
```
Comment (plbossart) [inline sound/soc/sof/control.c]: ah crap, this conflicts with the change I just merged, so you'll have to rebase and resubmit.

File: sound/soc/sof/control.c [hunk 3]
```diff
@@ -365,23 +385,10 @@ int snd_sof_bytes_ext_get(struct snd_kcontrol *kcontrol,
...
-
-	/* Prevent read of other kernel data */
-	if (size > max_length) {
-		dev_err(sdev->dev, "error: size is too large\n");
-		ret = -EINVAL;
-		goto out;
-	}
+	dev_dbg(sdev->dev, "bytes: user data size limit is %d\n", size);
+	dev_dbg(sdev->dev, "bytes: control bytes max size is %d\n", be->max);
+	dev_dbg(sdev->dev, "bytes: abi user data max size is %d\n", max_size);
```
Comment (plbossart) [inline sound/soc/sof/control.c]: same, it's too verbose and not very useful info

File: sound/soc/sof/topology.c [hunk 4]
```diff
@@ -316,20 +318,43 @@ static int sof_control_load_bytes(struct snd_soc_component *scomp,
...
-		cdata->data->abi = SOF_ABI_VERSION;
-	}
 
+		dev_dbg(sdev->dev, "tplg: bytes magic 0x%08x\n",
+			cdata->data->magic);
+		dev_dbg(sdev->dev, "tplg: bytes abi   0x%08x\n",
+			cdata->data->abi);
+		dev_dbg(sdev->dev, "tplg: bytes size  %d\n", cdata->data->size);
+		dev_dbg(sdev->dev, "tplg: priv size   %d\n",
+			control->priv.size);
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: too verbose?

**Comment by singalsu** [comment]
OK, I'll reduce the verbosity. I added them to keep my mind clear what the various sizes are while checking the size checks code. The error prints should now contain the sufficient information for a user to fix their data if there's issue.

**Comment by singalsu** [comment]
@plbossart I've updated now the PR.

**Comment by singalsu** [comment]
@plbossart @lgirdwood Great, thanks for the review rounds for this!

**Pull request merged**